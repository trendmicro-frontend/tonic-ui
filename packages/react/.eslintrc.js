module.exports = {
  extends: 'trendmicro',
  parser: '@babel/eslint-parser',
  env: {
    browser: true,
    node: true,
    jest: true,
  },
  plugins: [
    '@babel',
    'react-hooks',
  ],
  rules: {
    'react/prop-types': 0,
    'react-hooks/rules-of-hooks': 'error', // Checks rules of Hooks
    'react-hooks/exhaustive-deps': 'error', // Checks effect dependencies
  },
};
