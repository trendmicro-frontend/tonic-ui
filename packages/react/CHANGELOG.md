# @tonic-ui/react

## 2.0.3

### Patch Changes

- fix(react/OverflowTooltip): resolve tooltip misalignment issue with popup menu items by [@tinaClin](https://github.com/tinaClin) in [#894](https://github.com/trendmicro-frontend/tonic-ui/pull/894)

- feat(react): improve tooltip placement and offset calculations by [@cheton](https://github.com/cheton) in [#897](https://github.com/trendmicro-frontend/tonic-ui/pull/897)
