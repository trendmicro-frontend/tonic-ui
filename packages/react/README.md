React UI components built with Emotion and Styled System
