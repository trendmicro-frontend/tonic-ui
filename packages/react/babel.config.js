module.exports = {
  extends: '@trendmicro/babel-config',
  presets: [
    '@babel/preset-env',
    ['@babel/preset-react', {
      /**
       * Decides which runtime to use.
       * - 'automatic' auto imports the functions that JSX transpiles to.
       * - 'classic' does not automatic import anything.
       */
      'runtime': 'classic', // defaults to classic
      //'runtime': 'automatic', // for React 16.14.0, React 17 and higher that supports the new transform
    }],
    '@emotion/babel-preset-css-prop',
  ],
};
