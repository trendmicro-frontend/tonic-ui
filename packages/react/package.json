{
  "name": "@tonic-ui/react",
  "version": "1.0.0-alpha",
  "description": "React Tonic UI component library.",
  "main": "dist/index.js",
  "module": "dist/es/index.js",
  "publishConfig": {
    "access": "public",
    "directory": "build"
  },
  "scripts": {
    "build": "cross-env PACKAGE_NAME=index INPUT=src OUTPUT_DIRECTORY=dist rollup --config rollup.config.js",
    "clean": "del build coverage dist",
    "eslint": "eslint --ext .js,.jsx,.mjs .",
    "postbuild": "cpy --parents {dist/**,LICENSE,README.md,.npmignore,package.json} build",
    "pre-push": "yarn run eslint",
    "prepublish": "yarn run build",
    "test": "jest"
  },
  "dependencies": {
    "@emotion/is-prop-valid": "1.x",
    "@emotion/react": "11.x",
    "@emotion/styled": "11.x",
    "@popperjs/core": "2.x",
    "@tonic-ui/react-hooks": "^1.0.0-alpha",
    "@tonic-ui/styled-system": "^1.0.0-alpha",
    "@tonic-ui/theme": "^1.0.0-alpha",
    "@trendmicro/tmicon": "^1.21.0",
    "chained-function": "^0.5.0",
    "color": "3.x",
    "ensure-type": "1.x",
    "lodash.get": "4.x",
    "micro-memoize": "4.x",
    "prop-types": "15.x",
    "react-focus-lock": "2.x",
    "react-is": "^16.8.0 || ^17",
    "react-transition-group": "4.x"
  },
  "devDependencies": {
    "@babel/cli": "^7.0.0",
    "@babel/core": "^7.0.0",
    "@babel/eslint-parser": "^7.0.0",
    "@babel/eslint-plugin": "^7.0.0",
    "@babel/plugin-transform-runtime": "^7.0.0",
    "@babel/preset-env": "^7.0.0",
    "@babel/preset-react": "^7.0.0",
    "@emotion/babel-preset-css-prop": "~11.2.0",
    "@emotion/jest": "~11.5.0",
    "@rollup/plugin-node-resolve": "~11.2.0",
    "@testing-library/jest-dom": "^5.16.1",
    "@testing-library/react": "^12.1.2",
    "@testing-library/react-hooks": "^7.0.2",
    "@testing-library/user-event": "^13.5.0",
    "@trendmicro/babel-config": "^1.0.2",
    "babel-loader": "~8.2.3",
    "cpy-cli": "~3.1.1",
    "cross-env": "~7.0.3",
    "del-cli": "~4.0.1",
    "eslint": "^7.0.0",
    "eslint-config-trendmicro": "^2.1.0",
    "eslint-loader": "~4.0.2",
    "eslint-plugin-import": "latest",
    "eslint-plugin-jsx-a11y": "latest",
    "eslint-plugin-react": "latest",
    "eslint-plugin-react-hooks": "latest",
    "jest": "~27.3.1",
    "jest-extended": "^1.1.0",
    "react": "^16.8.0 || ^17",
    "react-dom": "^16.8.0 || ^17",
    "react-test-renderer": "^16.8.0 || ^17",
    "rollup": "~2.42.3",
    "rollup-plugin-babel": "~4.4.0"
  },
  "peerDependencies": {
    "react": "^16.8.0 || ^17",
    "react-dom": "^16.8.0 || ^17"
  },
  "jest": {
    "collectCoverage": true,
    "collectCoverageFrom": [
      "<rootDir>/src/**/*.js"
    ],
    "coverageReporters": [
      "lcov",
      "text",
      "html"
    ],
    "setupFiles": [],
    "setupFilesAfterEnv": [
      "<rootDir>/jest-setup.js"
    ],
    "testEnvironment": "jsdom",
    "testMatch": [
      "<rootDir>/src/**/*.test.js",
      "<rootDir>/test/**/*.js"
    ]
  },
  "author": "Cheton Wu <cheton_wu@trendmicro.com>",
  "homepage": "https://github.com/trendmicro-frontend/tonic-ui",
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/trendmicro-frontend/tonic-ui.git",
    "directory": "packages/react"
  },
  "bugs": {
    "url": "https://github.com/trendmicro-frontend/tonic-ui/issues"
  },
  "keywords": [
    "components",
    "emotion",
    "react",
    "styled-system",
    "tonic-ui"
  ]
}
