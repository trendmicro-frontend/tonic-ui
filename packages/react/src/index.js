// components
export * from './accordion';
export * from './alert';
export * from './badge';
export * from './box';
export * from './button';
export * from './checkbox';
export * from './color-mode';
export * from './color-style';
export * from './css-baseline';
export * from './divider';
export * from './drawer';
export * from './flat-button';
export * from './flex';
export * from './grid';
export * from './icon';
export * from './image';
export * from './input';
export * from './link';
export * from './menu';
export * from './modal';
export * from './pagination';
export * from './popover';
export * from './popper';
export * from './portal';
export * from './presence';
export * from './radio';
export * from './search-input';
export * from './scrollbar';
export * from './select';
export * from './skeleton';
export * from './space';
export * from './spinner';
export * from './stack';
export * from './switch';
export * from './table';
export * from './tabs';
export * from './tag';
export * from './text';
export * from './textarea';
export * from './theme';
export * from './toast';
export * from './tooltip';
export * from './transitions';
export * from './visually-hidden';

// hooks
export useDisclosure from './deprecated/useDisclosure'; // deprecated

// settings
export colorStyle from './shared/color-style';
export theme from './shared/theme';
