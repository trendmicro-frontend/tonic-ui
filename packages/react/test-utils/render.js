import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import { toHaveNoViolations } from 'jest-axe';
import * as React from 'react';
import { TonicProvider, theme } from '../src';

expect.extend(toHaveNoViolations);

const customRender = (ui, options) => {
  let wrapper = TonicProvider;
  const useCSSVariables = options?.useCSSVariables;
  if (useCSSVariables) {
    wrapper = ({ children }) => (
      <TonicProvider
        theme={{
          ...theme,
          config: {
            ...theme.config,
            useCSSVariables: true,
          },
        }}
      >
        {children}
      </TonicProvider>
    );
  }

  return render(ui, { wrapper, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
