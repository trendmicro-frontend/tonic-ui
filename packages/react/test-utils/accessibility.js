import { act } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import * as React from 'react';
import { render } from './render';

expect.extend(toHaveNoViolations);

const testA11y = async (ui, options) => {
  const { axeOptions, ...rest } = { ...options };
  const container = React.isValidElement(ui)
    ? render(ui, rest).container
    : ui;
  await act(async () => {
    const results = await axe(container, axeOptions);
    expect(results).toHaveNoViolations();
  });
};

export {
  testA11y,
};
