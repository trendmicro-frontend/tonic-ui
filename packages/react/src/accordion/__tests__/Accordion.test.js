import { screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { render } from '@tonic-ui/react/test-utils/render';
import { testA11y } from '@tonic-ui/react/test-utils/accessibility';
import {
  Accordion,
  AccordionItem,
  AccordionHeader,
  AccordionBody,
  Box,
  Text,
} from '@tonic-ui/react/src';
import React from 'react';

describe('Accordion', () => {
  it('should render correctly', async () => {
    const renderOptions = {
      useCSSVariables: true,
    };
    const { container } = render((
      <Accordion rowGap={1}>
        <AccordionItem defaultIsExpanded>
          <AccordionHeader>
            <Text>
              Accordion 1
            </Text>
          </AccordionHeader>
          <AccordionBody>
            <Box px="4x" py="3x" />
          </AccordionBody>
        </AccordionItem>
        <AccordionItem>
          <AccordionHeader>
            <Text>
              Accordion 2
            </Text>
          </AccordionHeader>
          <AccordionBody>
            <Box px="4x" py="3x" />
          </AccordionBody>
        </AccordionItem>
      </Accordion>
    ), renderOptions);

    expect(container).toMatchSnapshot();

    await testA11y(container, {
      axeOptions: {
        rules: {
          label: { enabled: false },
        },
      },
    });
  });

  it('should toggle visibility of accordion items when clicked', async () => {
    const user = userEvent.setup();
    render(
      <Accordion>
        <AccordionItem>
          <AccordionHeader data-testid="header-1">Section 1</AccordionHeader>
          <AccordionBody data-testid="body-1">Content 1</AccordionBody>
        </AccordionItem>
        <AccordionItem>
          <AccordionHeader data-testid="header-2">Section 2</AccordionHeader>
          <AccordionBody data-testid="body-2">Content 2</AccordionBody>
        </AccordionItem>
      </Accordion>
    );

    const header1 = screen.getByTestId('header-1');
    const body1 = screen.getByTestId('body-1');
    const header2 = screen.getByTestId('header-2');
    const body2 = screen.getByTestId('body-2');

    expect(header1).not.toHaveAttribute('aria-expanded');
    expect(header2).not.toHaveAttribute('aria-expanded');
    expect(body1).toHaveAttribute('aria-hidden', 'true');
    expect(body2).toHaveAttribute('aria-hidden', 'true');
    expect(body1).not.toBeVisible();
    expect(body2).not.toBeVisible();

    // Toggle the first accordion item
    await user.click(header1);
    expect(header1).toHaveAttribute('aria-expanded', 'true');
    expect(body1).not.toHaveAttribute('aria-hidden');
    expect(body1).toBeVisible();
    expect(body2).not.toBeVisible();

    // Toggle the second accordion item
    await user.click(header2);
    expect(header2).toHaveAttribute('aria-expanded', 'true');
    expect(body2).not.toHaveAttribute('aria-hidden');
    expect(body1).toBeVisible();
    expect(body2).toBeVisible();
  });

  it('should handle multiple accordion items independently', async () => {
    const user = userEvent.setup();
    const TestComponent = () => {
      const [expandedItem, setExpandedItem] = React.useState('item1');
      const handleToggle = item => ({ isExpanded }) => {
        setExpandedItem(isExpanded ? item : null);
      };

      return (
        <Accordion rowGap={1}>
          <AccordionItem
            isExpanded={expandedItem === 'item1'}
            onToggle={handleToggle('item1')}
          >
            <AccordionHeader data-testid="header-1">
              <Text>
                Accordion 1
              </Text>
            </AccordionHeader>
            <AccordionBody data-testid="body-1">
              <Box px="4x" py="3x" />
            </AccordionBody>
          </AccordionItem>
          <AccordionItem
            isExpanded={expandedItem === 'item2'}
            onToggle={handleToggle('item2')}
          >
            <AccordionHeader data-testid="header-2">
              <Text>
                Accordion 2
              </Text>
            </AccordionHeader>
            <AccordionBody data-testid="body-2">
              <Box px="4x" py="3x" />
            </AccordionBody>
          </AccordionItem>
        </Accordion>
      );
    };

    render(<TestComponent />);

    const header1 = screen.getByTestId('header-1');
    const body1 = screen.getByTestId('body-1');
    const header2 = screen.getByTestId('header-2');
    const body2 = screen.getByTestId('body-2');

    expect(header1).toHaveAttribute('aria-expanded', 'true');
    expect(header2).not.toHaveAttribute('aria-expanded');
    expect(body1).not.toHaveAttribute('aria-hidden');
    expect(body2).toHaveAttribute('aria-hidden', 'true');
    expect(body1).toBeVisible();
    expect(body2).not.toBeVisible();

    // Toggle the second accordion item
    await user.click(header2);
    expect(header1).not.toHaveAttribute('aria-expanded');
    expect(header2).toHaveAttribute('aria-expanded', 'true');
    expect(body1).toHaveAttribute('aria-hidden', 'true');
    expect(body2).not.toHaveAttribute('aria-hidden');
    expect(body1).not.toBeVisible();
    expect(body2).toBeVisible();
    expect(body2).toBeVisible();

    // Toggle the first accordion item
    await user.click(header1);
    expect(header1).toHaveAttribute('aria-expanded', 'true');
    expect(header2).not.toHaveAttribute('aria-expanded');
    expect(body1).not.toHaveAttribute('aria-hidden');
    expect(body2).toHaveAttribute('aria-hidden', 'true');
    expect(body1).toBeVisible();
    expect(body2).not.toBeVisible();
  });

  it('should render correctly with disabled attribute', () => {
    render(
      <AccordionItem disabled>
        <AccordionHeader data-testid="accordion-header">
          <Text>
            Disabled Accordion
          </Text>
        </AccordionHeader>
        <AccordionBody>
          <Box px="4x" py="3x" />
        </AccordionBody>
      </AccordionItem>
    );

    const accordionHeader = screen.getByTestId('accordion-header');
    expect(accordionHeader).toBeDisabled();
  });
});
