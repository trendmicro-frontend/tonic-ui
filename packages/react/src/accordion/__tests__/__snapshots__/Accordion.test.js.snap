// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render correctly 1`] = `
.emotion-0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  row-gap: 1px;
}

.emotion-4 {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background-color: var(--tonic-colors-gray-90);
  border: none;
  color: var(--tonic-colors-white-primary);
  cursor: pointer;
  outline: 0;
  padding: 0;
  width: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-column-gap: var(--tonic-sizes-4x);
  column-gap: var(--tonic-sizes-4x);
  border-color: var(--tonic-colors-transparent);
  border-style: solid;
  border-width: .0625rem;
  padding-left: 1rem;
  padding-right: 1rem;
  padding-top: calc(.75rem - .0625rem);
  padding-bottom: calc(.75rem - .0625rem);
}

.emotion-4:disabled,
.emotion-4:disabled:focus,
.emotion-4:disabled:hover,
.emotion-4[aria-disabled=true],
.emotion-4[aria-disabled=true]:focus,
.emotion-4[aria-disabled=true]:hover,
.emotion-4[data-disabled] {
  color: var(--tonic-colors-white-disabled);
}

.emotion-8 {
  display: block;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transform: rotate(-180deg);
  -moz-transform: rotate(-180deg);
  -ms-transform: rotate(-180deg);
  transform: rotate(-180deg);
  -webkit-transition: -webkit-transform 133ms cubic-bezier(0.0, 0, 0.2, 1);
  transition: transform 133ms cubic-bezier(0.0, 0, 0.2, 1);
}

.emotion-10:disabled,
.emotion-10:disabled:focus,
.emotion-10:disabled:hover,
.emotion-10[aria-disabled=true],
.emotion-10[aria-disabled=true]:focus,
.emotion-10[aria-disabled=true]:hover,
.emotion-10[data-disabled] {
  opacity: .28;
}

.emotion-12 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: var(--tonic-sizes-4x);
  height: var(--tonic-sizes-4x);
  vertical-align: middle;
}

.emotion-14 {
  height: auto;
  opacity: 1;
  -webkit-transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1),opacity 300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1),opacity 300ms cubic-bezier(0.4, 0, 0.2, 1);
}

.emotion-18 {
  padding-left: var(--tonic-space-4x);
  padding-right: var(--tonic-space-4x);
  padding-top: var(--tonic-space-3x);
  padding-bottom: var(--tonic-space-3x);
}

.emotion-28 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transform: rotate(0deg);
  -moz-transform: rotate(0deg);
  -ms-transform: rotate(0deg);
  transform: rotate(0deg);
  -webkit-transition: -webkit-transform 93ms cubic-bezier(0.0, 0, 0.2, 1);
  transition: transform 93ms cubic-bezier(0.0, 0, 0.2, 1);
}

.emotion-28:disabled,
.emotion-28:disabled:focus,
.emotion-28:disabled:hover,
.emotion-28[aria-disabled=true],
.emotion-28[aria-disabled=true]:focus,
.emotion-28[aria-disabled=true]:hover,
.emotion-28[data-disabled] {
  opacity: .28;
}

.emotion-32 {
  height: 0;
  opacity: 0;
  -webkit-transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1),opacity 300ms cubic-bezier(0.4, 0, 0.2, 1);
  transition: height 300ms cubic-bezier(0.4, 0, 0.2, 1),opacity 300ms cubic-bezier(0.4, 0, 0.2, 1);
  visibility: hidden;
}

<div>
  <div
    class="emotion-0 emotion-1"
  >
    <div
      class="emotion-2 emotion-1"
    >
      <button
        aria-controls="@tonic-ui/react:AccordionContent-1"
        aria-expanded="true"
        class="emotion-4 emotion-1"
        id="@tonic-ui/react:AccordionToggle-1"
        role="button"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-2 emotion-1"
        >
          <div
            class="emotion-8 emotion-1"
          >
            Accordion 1
          </div>
        </div>
        <div
          class="emotion-10 emotion-1"
        >
          <svg
            aria-hidden="true"
            class="emotion-12 emotion-1"
            data-icon="ChevronDownIcon"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 16 16"
          >
            <path
              d="M8 12l-7-6.607 1.476-1.393 5.524 5.214 5.524-5.214 1.476 1.393z"
            />
          </svg>
        </div>
      </button>
      <div
        aria-labelledby="@tonic-ui/react:AccordionToggle-1"
        class="emotion-14 emotion-1"
        id="@tonic-ui/react:AccordionContent-1"
        role="region"
      >
        <div
          class="emotion-2 emotion-1"
        >
          <div
            class="emotion-18 emotion-1"
          />
        </div>
      </div>
    </div>
    <div
      class="emotion-2 emotion-1"
    >
      <button
        aria-controls="@tonic-ui/react:AccordionContent-2"
        class="emotion-4 emotion-1"
        id="@tonic-ui/react:AccordionToggle-2"
        role="button"
        tabindex="0"
        type="button"
      >
        <div
          class="emotion-2 emotion-1"
        >
          <div
            class="emotion-8 emotion-1"
          >
            Accordion 2
          </div>
        </div>
        <div
          class="emotion-28 emotion-1"
        >
          <svg
            aria-hidden="true"
            class="emotion-12 emotion-1"
            data-icon="ChevronDownIcon"
            fill="currentColor"
            focusable="false"
            viewBox="0 0 16 16"
          >
            <path
              d="M8 12l-7-6.607 1.476-1.393 5.524 5.214 5.524-5.214 1.476 1.393z"
            />
          </svg>
        </div>
      </button>
      <div
        aria-hidden="true"
        aria-labelledby="@tonic-ui/react:AccordionToggle-2"
        class="emotion-32 emotion-1"
        id="@tonic-ui/react:AccordionContent-2"
        role="region"
      >
        <div
          class="emotion-2 emotion-1"
        >
          <div
            class="emotion-18 emotion-1"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
